File: Design-Patterns-Using-CSharp/Observer/PubSub.cs
using System;
using System.Collections.Generic;

class Publisher
{
    private Dictionary<string, List<Subscriber>> topicSubscribers = new Dictionary<string, List<Subscriber>>();

    public void Subscribe(Subscriber subscriber, string topic)
    {
        topicSubscribers.TryGetValue(topic, out var subscribers);
        if (subscribers == null)
        {
            subscribers = new List<Subscriber>();
            topicSubscribers[topic] = subscribers;
        }

        subscribers.Add(subscriber);
        Console.WriteLine($"Subscribing: {subscriber.GetId()} to topic: {topic}");
    }

    public void Unsubscribe(Subscriber subscriber, string topic)
    {
        if (topicSubscribers.TryGetValue(topic, out var subscribers))
        {
            subscribers.Remove(subscriber);
            Console.WriteLine($"Unsubscribing: {subscriber.GetId()} to topic: {topic}");
        }
    }

    public void NotifySubscribers(string data, string topic)
    {
        if (topicSubscribers.TryGetValue(topic, out var subscribers))
        {
            Console.WriteLine($"Publishing: {data} in topic: {topic}");
            foreach (var subscriber in subscribers)
            {
                subscriber.Update(data);
            }
        }
    }
}

class Subscriber
{
    private string id;

    public Subscriber(string id)
    {
        this.id = id;
    }

    public string GetId()
    {
        return id;
    }

    public void Update(string data)
    {
        Console.WriteLine($"Subscriber {id} got :: {data}");
    }
}

// Client code

class PubSub
{
    static void Main(string[] args)
    {
        Publisher pub = new Publisher();

        Subscriber sub1 = new Subscriber("Subscriber1");
        Subscriber sub2 = new Subscriber("Subscriber2");
        Subscriber sub3 = new Subscriber("Subscriber3");

        Console.WriteLine();
        pub.Subscribe(sub1, "topic1");
        pub.Subscribe(sub2, "topic2");
        pub.Subscribe(sub3, "topic2");

        Console.WriteLine();
        pub.NotifySubscribers("Topic 1 data", "topic1");

        Console.WriteLine();
        pub.NotifySubscribers("Topic 2 data", "topic2");

        Console.WriteLine();
        pub.Unsubscribe(sub3, "topic2");
        pub.NotifySubscribers("Topic 2 data", "topic2");
    }
}

/*
Subscribing: Subscriber1 to topic: topic1
Subscribing: Subscriber2 to topic: topic2
Subscribing: Subscriber3 to topic: topic2

Publishing: Topic 1 data in topic: topic1
Subscriber Subscriber1 got :: Topic 1 data

Publishing: Topic 2 data in topic: topic2
Subscriber Subscriber2 got :: Topic 2 data
Subscriber Subscriber3 got :: Topic 2 data

Unsubscribing: Subscriber3 to topic: topic2
Publishing: Topic 2 data in topic: topic2
Subscriber Subscriber2 got :: Topic 2 data
*/

File: Design-Patterns-Using-Ruby/Observer/PubSub.rb
class Publisher
    attr_reader :topic_subscribers

    def initialize
        @topic_subscribers = Hash.new { |hash, key| hash[key] = [] }
    end

    def subscribe(subscriber, topic)
        @topic_subscribers[topic] << subscriber
        puts "Subscribing: #{subscriber.id} to topic: #{topic}"
    end

    def unsubscribe(subscriber, topic)
        @topic_subscribers[topic].delete(subscriber)
        puts "Unsubscribing: #{subscriber.id} to topic: #{topic}"
    end

    def notify_subscribers(data, topic)
        if @topic_subscribers.key?(topic)
            puts "Publishing: #{data} in topic: #{topic}"
            @topic_subscribers[topic].each { |subscriber| subscriber.update(data) }
        end
    end
end

class Subscriber
    attr_reader :id

    def initialize(id)
        @id = id
    end

    def update(data)
        puts "Subscriber #{id} got :: #{data}"
    end
end

# Client code
pub = Publisher.new

sub1 = Subscriber.new("Subscriber1")
sub2 = Subscriber.new("Subscriber2")
sub3 = Subscriber.new("Subscriber3")

puts
pub.subscribe(sub1, "topic1")
pub.subscribe(sub2, "topic2")
pub.subscribe(sub3, "topic2")

puts
pub.notify_subscribers("Topic 1 data", "topic1")

puts
pub.notify_subscribers("Topic 2 data", "topic2")

puts
pub.unsubscribe(sub3, "topic2")
pub.notify_subscribers("Topic 2 data", "topic2")

=begin 
Subscribing: Subscriber1 to topic: topic1
Subscribing: Subscriber2 to topic: topic2
Subscribing: Subscriber3 to topic: topic2

Publishing: Topic 1 data in topic: topic1
Subscriber Subscriber1 got :: Topic 1 data

Publishing: Topic 2 data in topic: topic2
Subscriber Subscriber2 got :: Topic 2 data
Subscriber Subscriber3 got :: Topic 2 data

Unsubscribing: Subscriber3 to topic: topic2
Publishing: Topic 2 data in topic: topic2
Subscriber Subscriber2 got :: Topic 2 data 
=end